common/include/pcl/experimental/executor/best_fit.hpp

```diff
@@ -0,0 +1,158 @@
+    return true;
+  }
+
+  template <typename Executor,
+            typename executor::InstanceOf<Executor, executor::omp_executor> = 0>
+  static bool check(Executor& exec) {
+    if (const char* env_p = std::getenv("PCL_ENABLE_OMP_EXEC"))
+      return strcasecmp(env_p, "OFF") == 0;
+
+    // If hardware_concurrency() fails to get the number of threads than set max
+    // threads to 2 as a fallback to prevent unwanted scaling in machines with
+    // large number of available threads
+    auto max_threads =
+        std::max(std::min(std::thread::hardware_concurrency(), 8u), 2u);
+    exec.set_max_threads(max_threads);
+
+    return true;
+  }
+
+  template <typename Executor, typename executor::InstanceOf<
+                                   Executor, executor::cuda_executor> = 0>
+  static bool check(Executor& exec) {
+    pcl::utils::ignore(exec);
+    if (const char* env_p = std::getenv("PCL_ENABLE_CUDA_EXEC"))
+      return strcasecmp(env_p, "ON") == 0;
+    return true;
+  }
+};
+
+namespace detail {
+
+template <typename Function, typename Executor, typename = void>
+bool execute(Function&& f, Executor& exec) {
+  pcl::utils::ignore(f, exec);
+  return false;
+}
+
+template <typename Function, template <typename...> class Executor,
+          typename... Properties,
+          typename std::enable_if<executor::is_executor_available_v<Executor>,
+                                  int>::type = 0>
+bool execute(Function&& f, Executor<Properties...>& exec) {
+  f(exec);
+  return true;
+}
+
+template <typename Supported>
+struct executor_predicate {
+  template <typename T, typename = void>
+  struct condition : std::false_type {};
+
+  template <typename T>
+  struct condition<T,
+                   std::enable_if_t<is_executor_instance_available<T>::value &&
+                                    tuple_contains_type<T, Supported>::value>>
+      : std::true_type {};
+};
+
+}  // namespace detail
+
+template <typename RuntimeChecks = executor_runtime_checks, typename Function,
+          typename... SupportedExecutors>
+void enable_exec_with_priority(
+    Function&& f, std::tuple<SupportedExecutors...> supported_execs) {
+  static_assert(std::is_base_of<executor_runtime_checks, RuntimeChecks>::value,
+                "Runtime checks should inherit from executor_runtime_checks");
+  bool executor_selected = false;
+  executor::for_each_until_true(supported_execs, [&](auto& exec) {
+    if (RuntimeChecks::check(exec)) {
+      executor_selected = detail::execute(f, exec);
+      return executor_selected;
+    }
+    return false;
+  });
+
+  // This should not happen, at least one executor should always be selected. So
+  // either all runtime checks failed which is incorrect behaviour or no
+  // supported executors were passed which should not be done
+  if (!executor_selected)
+    std::cerr << "No executor selected. All runtime checks returned false or "
+                 "no executors were passed."
+              << std::endl;
```

Comment:
Is `cerr` a strong enough reason?

Reply:
I don't know, will need to read more about the deep dark mysteries of `iostream`. The buffer might get flushed before a graceful exit (but not a forced exit for example). `cerr` might be unbuffered so this might be a moot discussion.

Reply:
Give the uncertain nature, I will leave it as is.